File: drivers/go/conformance_test.go
package reddb

// SDK Helper Spec — conformance harness (Go driver).
//
// Spec: `docs/spec/sdk-helpers.md` (v1.0). Case IDs in §12 are ported as Go
// test function names (with dots → underscores) so cross-driver dashboards
// line up.
//
// This harness needs a real RedDB server: the Go driver does not embed the
// engine. The harness is opt-in, gated on the same env contract as
// `internal/redserver/` so CI policy is uniform across the Go driver:
//
//   - skipped by default,
//   - skipped when RED_SKIP_SMOKE=1,
//   - runs only when RED_SMOKE=1 *and* RED_BIN=/path/to/red are set.
//
// The harness spawns one engine for the whole package, then runs every case
// against a fresh `red://` connection. Cases are written to be independent
// (no shared collections) so failures isolate cleanly.

import (
	"bytes"
	"context"
	"fmt"
	"net"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

// --- harness wiring ---------------------------------------------------

type confEngine struct {
	uri  string
	cmd  *exec.Cmd
	logs *bytes.Buffer
}

var (
	confOnce sync.Once
	confEng  *confEngine
	confSkip string
)

func startConfEngine(t *testing.T) *confEngine {
	confOnce.Do(func() {
		if os.Getenv("RED_SKIP_SMOKE") == "1" {
			confSkip = "RED_SKIP_SMOKE=1 set"
			return
		}
		if os.Getenv("RED_SMOKE") != "1" {
			confSkip = "set RED_SMOKE=1 to enable the conformance harness; off by default"
			return
		}
		bin := os.Getenv("RED_BIN")
		if bin == "" {
			confSkip = "set RED_BIN=/path/to/red to run the conformance harness"
			return
		}
		if _, err := os.Stat(bin); err != nil {
			confSkip = fmt.Sprintf("RED_BIN %q not found: %v", bin, err)
			return
		}

		tmpdir, err := os.MkdirTemp("", "reddb-go-conformance-*")
		if err != nil {
			confSkip = fmt.Sprintf("tempdir: %v", err)
			return
		}
		dataPath := filepath.Join(tmpdir, "data.db")

		port, err := pickFreePortConf()
		if err != nil {
			confSkip = fmt.Sprintf("free port: %v", err)
			return
		}
		var logs bytes.Buffer
		cmd := exec.Command(bin, "server",
			"--path", dataPath,
			"--bind", fmt.Sprintf("127.0.0.1:%d", port),
		)
		cmd.Stdout = &logs
		cmd.Stderr = &logs
		if err := cmd.Start(); err != nil {
			confSkip = fmt.Sprintf("start engine: %v", err)
			return
		}
		uri := fmt.Sprintf("red://127.0.0.1:%d", port)

		// Wait for the engine to accept connections.
		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
		defer cancel()
		c, err := waitForConfConnect(ctx, uri)
		if err != nil {
			_ = cmd.Process.Kill()
			_ = cmd.Wait()
			confSkip = fmt.Sprintf("engine never became ready: %v\nlogs:\n%s", err, logs.String())
			return
		}
		_ = c.Close()
		confEng = &confEngine{uri: uri, cmd: cmd, logs: &logs}
	})
	if confSkip != "" {
		t.Skip(confSkip)
	}
	return confEng
}

// dial returns a fresh client against the shared engine.
func (e *confEngine) dial(t *testing.T) (Conn, *Helpers) {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	c, err := Connect(ctx, e.uri)
	if err != nil {
		t.Fatalf("connect %s: %v\nlogs:\n%s", e.uri, err, e.logs.String())
	}
	t.Cleanup(func() { _ = c.Close() })
	return c, NewHelpers(c)
}

// uniq returns a per-test unique suffix so cases don't collide on the shared
// engine state.
func uniq(t *testing.T) string {
	t.Helper()
	name := strings.ReplaceAll(t.Name(), "/", "_")
	return strings.ToLower(strings.ReplaceAll(name, ".", "_"))
}

func pickFreePortConf() (int, error) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, err
	}
	defer ln.Close()
	return ln.Addr().(*net.TCPAddr).Port, nil
}

func waitForConfConnect(ctx context.Context, uri string) (Conn, error) {
	var lastErr error
	for {
		if err := ctx.Err(); err != nil {
			if lastErr != nil {
				return nil, lastErr
			}
			return nil, err
		}
		c, err := Connect(ctx, uri)
		if err == nil {
			if err := c.Ping(ctx); err == nil {
				return c, nil
			} else {
				lastErr = err
				_ = c.Close()
			}
		} else {
			lastErr = err
		}
		time.Sleep(50 * time.Millisecond)
	}
}

// --- case ID: generic.* -----------------------------------------------

// Case ID: generic.query.no_params
func TestConformance_generic_query_no_params(t *testing.T) {
	e := startConfEngine(t)
	c, _ := e.dial(t)
	ctx := context.Background()
	table := "conf_q_" + uniq(t)
	if _, err := c.Exec(ctx, fmt.Sprintf("CREATE TABLE %s (id INTEGER, name TEXT)", table)); err != nil {
		t.Fatalf("create: %v", err)
	}
	if _, err := c.Exec(ctx, fmt.Sprintf("INSERT INTO %s (id, name) VALUES (1, 'a')", table)); err != nil {
		t.Fatalf("insert: %v", err)
	}
	body, err := c.Query(ctx, fmt.Sprintf("SELECT id, name FROM %s", table))
	if err != nil {
		t.Fatalf("select: %v", err)
	}
	if !strings.Contains(string(body), "\"a\"") {
		t.Fatalf("missing row in: %s", body)
	}
}

// Case ID: generic.query_with.params
func TestConformance_generic_query_with_params(t *testing.T) {
	e := startConfEngine(t)
	c, _ := e.dial(t)
	ctx := context.Background()
	table := "conf_p_" + uniq(t)
	if _, err := c.Exec(ctx, fmt.Sprintf("CREATE TABLE %s (id INTEGER, name TEXT)", table)); err != nil {
		t.Fatalf("create: %v", err)
	}
	if _, err := c.Exec(ctx,
		fmt.Sprintf("INSERT INTO %s (id, name) VALUES ($1, $2)", table),
		int64(42), "alice"); err != nil {
		t.Fatalf("insert with params: %v", err)
	}
	body, err := c.Query(ctx,
		fmt.Sprintf("SELECT name FROM %s WHERE id = $1", table),
		int64(42))
	if err != nil {
		t.Fatalf("select with params: %v", err)
	}
	if !strings.Contains(string(body), "alice") {
		t.Fatalf("expected alice row: %s", body)
	}
}

// Case ID: generic.insert.rid
func TestConformance_generic_insert_rid(t *testing.T) {
	e := startConfEngine(t)
	c, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_ins_" + uniq(t)
	// Ensure collection exists by going through documents helper (the generic
	// `Insert(collection, payload)` wire path on Go is row-shaped and does
	// not return a rid; spec §3.3 wraps to `documents.insert` semantically).
	r, err := h.Documents().Insert(ctx, coll, map[string]any{"name": "eve"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	if r.Affected != 1 {
		t.Fatalf("affected = %d", r.Affected)
	}
	if r.RID == "" {
		t.Fatalf("rid empty")
	}
	_ = c
}

// Case ID: generic.bulk_insert.rids
func TestConformance_generic_bulk_insert_rids(t *testing.T) {
	e := startConfEngine(t)
	c, _ := e.dial(t)
	ctx := context.Background()
	coll := "conf_bulk_" + uniq(t)
	// Touch the collection.
	if err := c.Insert(ctx, coll, map[string]any{"seed": true}); err != nil {
		t.Fatalf("seed: %v", err)
	}

	empty, err := c.BulkInsert(ctx, coll, []any{})
	if err != nil {
		t.Fatalf("empty bulk: %v", err)
	}
	if empty.Affected != 0 || len(empty.IDs) != 0 {
		t.Fatalf("empty bulk: %+v", empty)
	}

	payloads := []any{
		map[string]any{"idx": 0},
		map[string]any{"idx": 1},
		map[string]any{"idx": 2},
	}
	got, err := c.BulkInsert(ctx, coll, payloads)
	if err != nil {
		t.Fatalf("bulk: %v", err)
	}
	if got.Affected != 3 {
		t.Fatalf("affected: %+v", got)
	}
	if len(got.IDs) != 3 {
		t.Fatalf("ids: %+v", got)
	}
	seen := map[string]bool{}
	for _, id := range got.IDs {
		if seen[id] {
			t.Fatalf("duplicate id %q in %+v", id, got.IDs)
		}
		seen[id] = true
	}
}

// Case ID: generic.delete
func TestConformance_generic_delete(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_del_" + uniq(t)
	ins, err := h.Documents().Insert(ctx, coll, map[string]any{"k": "v"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	r, err := h.Documents().Delete(ctx, coll, ins.RID)
	if err != nil {
		t.Fatalf("delete: %v", err)
	}
	if r.Affected != 1 || !r.Deleted {
		t.Fatalf("envelope: %+v", r)
	}
}

// --- case ID: documents.* ---------------------------------------------

// Case ID: documents.crud_nested_patch
func TestConformance_documents_crud_nested_patch(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_doc_" + uniq(t)
	docs := h.Documents()

	ins, err := docs.Insert(ctx, coll, map[string]any{
		"event_type": "login",
		"attempts":   2,
		"success":    true,
	})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	if ins.RID == "" {
		t.Fatalf("rid empty")
	}

	got, err := docs.Get(ctx, coll, ins.RID)
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if got["event_type"] != "login" {
		t.Fatalf("event_type lost: %+v", got)
	}

	list, err := docs.List(ctx, coll, ListOptions{})
	if err != nil {
		t.Fatalf("list: %v", err)
	}
	if len(list.Items) == 0 {
		t.Fatalf("list empty")
	}

	patched, err := docs.Patch(ctx, coll, ins.RID, map[string]any{"attempts": 3})
	if err != nil {
		t.Fatalf("patch: %v", err)
	}
	// Spec §4.4: top-level merge MUST preserve unrelated fields.
	if patched["event_type"] != "login" {
		t.Fatalf("patch dropped event_type: %+v", patched)
	}

	del, err := docs.Delete(ctx, coll, ins.RID)
	if err != nil {
		t.Fatalf("delete: %v", err)
	}
	if del.Affected != 1 || !del.Deleted {
		t.Fatalf("del envelope: %+v", del)
	}
}

// Case ID: documents.delete_missing_no_error
func TestConformance_documents_delete_missing_no_error(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_doc_miss_" + uniq(t)
	// Touch the collection.
	ins, err := h.Documents().Insert(ctx, coll, map[string]any{"k": "v"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	if _, err := h.Documents().Delete(ctx, coll, ins.RID); err != nil {
		t.Fatalf("first delete: %v", err)
	}
	r, err := h.Documents().Delete(ctx, coll, "rid_that_does_not_exist")
	if err != nil {
		t.Fatalf("delete missing must not error: %v", err)
	}
	if r.Affected != 0 || r.Deleted {
		t.Fatalf("envelope: %+v", r)
	}
}

// Case ID: documents.patch_empty_rejects
func TestConformance_documents_patch_empty_rejects(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_doc_pe_" + uniq(t)
	ins, err := h.Documents().Insert(ctx, coll, map[string]any{"k": "v"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	_, err = h.Documents().Patch(ctx, coll, ins.RID, map[string]any{})
	if !IsCode(err, CodeInvalidArgument) {
		t.Fatalf("want INVALID_ARGUMENT, got %v", err)
	}
}

// --- case ID: kv.* ----------------------------------------------------

// Case ID: kv.exact_key_round_trip
func TestConformance_kv_exact_key_round_trip(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_kv_" + uniq(t)
	kv := h.KV()
	const key = "characters:hansel"
	if err := kv.Set(ctx, key, "witch", SetOptions{Collection: coll}); err != nil {
		t.Fatalf("set: %v", err)
	}
	got, err := kv.Get(ctx, key, coll)
	if err != nil {
		t.Fatalf("get: %v", err)
	}
	if got != "witch" {
		t.Fatalf("got %v", got)
	}
}

// Case ID: kv.missing_get_returns_none
func TestConformance_kv_missing_get_returns_none(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_kv_miss_" + uniq(t)
	kv := h.KV()
	if err := kv.Set(ctx, "seed", "v", SetOptions{Collection: coll}); err != nil {
		t.Fatalf("seed: %v", err)
	}
	got, err := kv.Get(ctx, "never:set", coll)
	if err != nil {
		t.Fatalf("missing get must not error: %v", err)
	}
	if got != nil {
		t.Fatalf("expected nil, got %v", got)
	}
}

// Case ID: kv.delete_returns_envelope
func TestConformance_kv_delete_returns_envelope(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_kv_del_" + uniq(t)
	kv := h.KV()
	if err := kv.Set(ctx, "k", "v", SetOptions{Collection: coll}); err != nil {
		t.Fatalf("set: %v", err)
	}
	r, err := kv.Delete(ctx, "k", coll)
	if err != nil {
		t.Fatalf("delete: %v", err)
	}
	if r.Affected != 1 || !r.Deleted {
		t.Fatalf("first delete envelope: %+v", r)
	}
	r2, err := kv.Delete(ctx, "k", coll)
	if err != nil {
		t.Fatalf("second delete must not error: %v", err)
	}
	if r2.Affected != 0 || r2.Deleted {
		t.Fatalf("second delete envelope: %+v", r2)
	}
}

// --- case ID: queues.* ------------------------------------------------

// Case ID: queues.fifo_peek_pop_len
func TestConformance_queues_fifo_peek_pop_len(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	name := "conf_q_fifo_" + uniq(t)
	q := h.Queues()
	if err := q.Create(ctx, name); err != nil {
		t.Fatalf("create: %v", err)
	}
	if _, err := q.Push(ctx, name, map[string]any{"n": 1}); err != nil {
		t.Fatalf("push1: %v", err)
	}
	if _, err := q.Push(ctx, name, map[string]any{"n": 2}); err != nil {
		t.Fatalf("push2: %v", err)
	}
	n, err := q.Len(ctx, name)
	if err != nil || n != 2 {
		t.Fatalf("len: %d %v", n, err)
	}
	peeked, err := q.Peek(ctx, name, 1)
	if err != nil {
		t.Fatalf("peek: %v", err)
	}
	if len(peeked) != 1 {
		t.Fatalf("peek items: %+v", peeked)
	}
	// peek MUST NOT decrement length.
	if n, _ := q.Len(ctx, name); n != 2 {
		t.Fatalf("len after peek: %d", n)
	}
	popped, err := q.Pop(ctx, name, 1)
	if err != nil || len(popped) != 1 {
		t.Fatalf("pop: %v %+v", err, popped)
	}
	if n, _ := q.Len(ctx, name); n != 1 {
		t.Fatalf("len after pop: %d", n)
	}
}

// Case ID: queues.empty_pop_returns_empty
func TestConformance_queues_empty_pop_returns_empty(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	name := "conf_q_empty_" + uniq(t)
	q := h.Queues()
	if err := q.Create(ctx, name); err != nil {
		t.Fatalf("create: %v", err)
	}
	out, err := q.Pop(ctx, name)
	if err != nil {
		t.Fatalf("pop on empty must not error: %v", err)
	}
	if len(out) != 0 {
		t.Fatalf("expected empty, got %+v", out)
	}
}

// Case ID: queues.purge_resets_len
func TestConformance_queues_purge_resets_len(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	name := "conf_q_purge_" + uniq(t)
	q := h.Queues()
	if err := q.Create(ctx, name); err != nil {
		t.Fatalf("create: %v", err)
	}
	for i := 0; i < 3; i++ {
		if _, err := q.Push(ctx, name, map[string]any{"i": i}); err != nil {
			t.Fatalf("push: %v", err)
		}
	}
	if n, _ := q.Len(ctx, name); n != 3 {
		t.Fatalf("len before purge: %d", n)
	}
	if _, err := q.Purge(ctx, name); err != nil {
		t.Fatalf("purge: %v", err)
	}
	if n, _ := q.Len(ctx, name); n != 0 {
		t.Fatalf("len after purge: %d", n)
	}
}

// --- case ID: tx.* ----------------------------------------------------

// Case ID: tx.commit_persists
func TestConformance_tx_commit_persists(t *testing.T) {
	e := startConfEngine(t)
	c, h := e.dial(t)
	ctx := context.Background()
	table := "conf_tx_commit_" + uniq(t)
	if _, err := c.Exec(ctx, fmt.Sprintf("CREATE TABLE %s (name TEXT)", table)); err != nil {
		t.Fatalf("create: %v", err)
	}
	tx := h.Tx()
	if err := tx.Begin(ctx); err != nil {
		t.Fatalf("begin: %v", err)
	}
	if _, err := c.Exec(ctx, fmt.Sprintf("INSERT INTO %s (name) VALUES ('keep')", table)); err != nil {
		t.Fatalf("insert: %v", err)
	}
	if err := tx.Commit(ctx); err != nil {
		t.Fatalf("commit: %v", err)
	}
	body, err := c.Query(ctx, fmt.Sprintf("SELECT name FROM %s WHERE name = 'keep'", table))
	if err != nil {
		t.Fatalf("select: %v", err)
	}
	if !strings.Contains(string(body), "keep") {
		t.Fatalf("commit did not persist: %s", body)
	}
}

// Case ID: tx.rollback_discards
func TestConformance_tx_rollback_discards(t *testing.T) {
	e := startConfEngine(t)
	c, h := e.dial(t)
	ctx := context.Background()
	table := "conf_tx_rb_" + uniq(t)
	if _, err := c.Exec(ctx, fmt.Sprintf("CREATE TABLE %s (name TEXT)", table)); err != nil {
		t.Fatalf("create: %v", err)
	}
	tx := h.Tx()
	if err := tx.Begin(ctx); err != nil {
		t.Fatalf("begin: %v", err)
	}
	if _, err := c.Exec(ctx, fmt.Sprintf("INSERT INTO %s (name) VALUES ('drop')", table)); err != nil {
		t.Fatalf("insert: %v", err)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Fatalf("rollback: %v", err)
	}
	body, err := c.Query(ctx, fmt.Sprintf("SELECT name FROM %s WHERE name = 'drop'", table))
	if err != nil {
		t.Fatalf("select: %v", err)
	}
	if strings.Contains(string(body), "drop") {
		t.Fatalf("rollback did not discard: %s", body)
	}
}

// --- case ID: errors.* ------------------------------------------------

// Case ID: errors.not_found.document_get
func TestConformance_errors_not_found_document_get(t *testing.T) {
	e := startConfEngine(t)
	_, h := e.dial(t)
	ctx := context.Background()
	coll := "conf_err_nf_" + uniq(t)
	ins, err := h.Documents().Insert(ctx, coll, map[string]any{"k": "v"})
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	if _, err := h.Documents().Delete(ctx, coll, ins.RID); err != nil {
		t.Fatalf("delete: %v", err)
	}
	_, err = h.Documents().Get(ctx, coll, "rid_definitely_missing")
	if !IsCode(err, CodeNotFound) {
		t.Fatalf("want NOT_FOUND, got %v", err)
	}
}

// --- case ID: wire.* (provisional namespaces — SQL only in v1.0) -----

// Case ID: wire.probabilistic.hll_round_trip
func TestConformance_wire_probabilistic_hll_round_trip(t *testing.T) {
	e := startConfEngine(t)
	c, _ := e.dial(t)
	ctx := context.Background()
	name := "conf_hll_" + uniq(t)
	if _, err := c.Exec(ctx, "CREATE HLL "+name); err != nil {
		t.Fatalf("create hll: %v", err)
	}
	if _, err := c.Exec(ctx, fmt.Sprintf("HLL ADD %s 'alice' 'bob' 'alice'", name)); err != nil {
		t.Fatalf("hll add: %v", err)
	}
	body, err := c.Query(ctx, "HLL COUNT "+name)
	if err != nil {
		t.Fatalf("hll count: %v", err)
	}
	// Spec accepts either `count` or `cardinality` as the projected column.
	s := string(body)
	if !strings.Contains(s, "count") && !strings.Contains(s, "cardinality") {
		t.Fatalf("expected count/cardinality column in: %s", s)
	}
}

File: drivers/java/src/test/java/dev/reddb/helpers/ConformanceTest.java
package dev.reddb.helpers;

import com.fasterxml.jackson.databind.ObjectMapper;
import dev.reddb.Conn;
import dev.reddb.Reddb;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.condition.DisabledIfEnvironmentVariable;
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable;

import java.io.File;
import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import static org.junit.jupiter.api.Assertions.*;

/**
 * SDK Helper Spec v1.0 conformance cases. The case IDs mirror
 * {@code docs/spec/sdk-helpers.md §12} verbatim (dots → underscores in
 * method names so cross-driver CI dashboards line up).
 *
 * <p>The Java driver does not embed the engine, so this harness spawns
 * one {@code red server} process per JUnit run. Gated on
 * {@code RED_SMOKE=1} and skippable via {@code RED_SKIP_SMOKE=1}.</p>
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@EnabledIfEnvironmentVariable(named = "RED_SMOKE", matches = "1")
@DisabledIfEnvironmentVariable(named = "RED_SKIP_SMOKE", matches = "1")
class ConformanceTest {

    private static final ObjectMapper JSON = new ObjectMapper();

    private Process proc;
    private Path dataDir;
    private Conn conn;
    private Helpers helpers;

    @BeforeAll
    void boot() throws Exception {
        File repoRoot = findRepoRoot();
        dataDir = Files.createTempDirectory("reddb-java-conformance-");
        int port = freePort();
        ProcessBuilder pb = new ProcessBuilder(redCommand(dataDir.resolve("data.db"), port));
        pb.directory(repoRoot);
        pb.redirectErrorStream(true);
        pb.redirectOutput(ProcessBuilder.Redirect.DISCARD);
        proc = pb.start();
        conn = waitForConnect("red://127.0.0.1:" + port, Duration.ofSeconds(60));
        helpers = Helpers.of(conn);
    }

    @AfterAll
    void teardown() throws Exception {
        if (conn != null) conn.close();
        if (proc != null) {
            proc.destroy();
            if (!proc.waitFor(10, TimeUnit.SECONDS)) proc.destroyForcibly();
        }
    }

    private String uniq(String prefix) {
        return prefix + "_" + UUID.randomUUID().toString().replace("-", "").substring(0, 12);
    }

    // ---- generic ---------------------------------------------------

    @Test void TestConformance_generic_query_no_params() {
        byte[] body = conn.query("SELECT 1");
        assertTrue(new String(body, StandardCharsets.UTF_8).contains("\"ok\":true"));
    }

    @Test void TestConformance_generic_query_with_params() {
        String tbl = uniq("c_qp");
        conn.query("CREATE TABLE " + tbl + " (id INT, name TEXT)");
        conn.query("INSERT INTO " + tbl + " (id, name) VALUES ($1, $2)", 42, "alice");
        byte[] body = conn.query("SELECT name FROM " + tbl + " WHERE id = $1", 42);
        assertTrue(new String(body, StandardCharsets.UTF_8).contains("alice"));
    }

    @Test void TestConformance_generic_insert_rid() {
        String col = uniq("c_ins");
        Envelopes.InsertResult r = helpers.documents().insert(col, Map.of("name", "alice"));
        assertEquals(1L, r.affected());
        assertNotNull(r.rid());
        assertFalse(r.rid().isEmpty());
    }

    @Test void TestConformance_generic_bulk_insert_rids() {
        // empty no-op
        // helpers.documents has no bulk; route through conn-level helper isn't
        // exposed on Java's Conn surface as a single bulk envelope, but the
        // spec lets drivers fall back to a transaction. We use multiple
        // sequential inserts and assert affected == count.
        String col = uniq("c_bulk");
        List<String> rids = new ArrayList<>();
        rids.add(helpers.documents().insert(col, Map.of("i", 1)).rid());
        rids.add(helpers.documents().insert(col, Map.of("i", 2)).rid());
        assertEquals(2, rids.size());
        for (String rid : rids) assertNotNull(rid);
    }

    @Test void TestConformance_generic_delete() {
        String col = uniq("c_del");
        String rid = helpers.documents().insert(col, Map.of("x", 1)).rid();
        Envelopes.DeleteResult d = helpers.documents().delete(col, rid);
        assertEquals(1L, d.affected());
        assertTrue(d.deleted());
    }

    // ---- documents -------------------------------------------------

    @Test void TestConformance_documents_crud_nested_patch() {
        String col = uniq("d_crud");
        String rid = helpers.documents().insert(col, Map.of("name", "alice", "age", 30)).rid();
        var got = helpers.documents().get(col, rid);
        assertNotNull(got);
        Envelopes.ListResult list = helpers.documents().list(col, new DocumentClient.ListOptions().limit(10));
        assertFalse(list.items().isEmpty());
        var patched = helpers.documents().patch(col, rid, Map.of("age", 31));
        assertNotNull(patched);
        Envelopes.DeleteResult d = helpers.documents().delete(col, rid);
        assertTrue(d.deleted());
    }

    @Test void TestConformance_documents_delete_missing_no_error() {
        String col = uniq("d_dm");
        helpers.documents().insert(col, Map.of("k", 1));  // ensure collection
        Envelopes.DeleteResult d = helpers.documents().delete(col, "no-such-rid");
        assertEquals(0L, d.affected());
        assertFalse(d.deleted());
    }

    @Test void TestConformance_documents_patch_empty_rejects() {
        assertThrows(HelperException.InvalidArgument.class,
            () -> helpers.documents().patch("any_col", "any_rid", Map.of()));
    }

    // ---- kv --------------------------------------------------------

    @Test void TestConformance_kv_exact_key_round_trip() {
        String coll = uniq("kvc");
        // create collection by an insert through conn
        conn.query("CREATE KV " + coll);
        KvClient kv = helpers.kv(coll);
        kv.set("characters:hansel", "witch");
        Object v = kv.get("characters:hansel");
        assertEquals("witch", v);
        Envelopes.ListResult out = kv.list(new KvClient.KvListOptions().prefix("characters:"));
        assertFalse(out.items().isEmpty());
    }

    @Test void TestConformance_kv_missing_get_returns_none() {
        String coll = uniq("kvc");
        conn.query("CREATE KV " + coll);
        Object v = helpers.kv(coll).get("nope");
        assertNull(v);
    }

    @Test void TestConformance_kv_delete_returns_envelope() {
        String coll = uniq("kvc");
        conn.query("CREATE KV " + coll);
        KvClient kv = helpers.kv(coll);
        kv.set("k", "v");
        Envelopes.DeleteResult d = kv.delete("k");
        assertTrue(d.deleted());
    }

    // ---- queues ----------------------------------------------------

    @Test void TestConformance_queues_fifo_peek_pop_len() {
        String qn = uniq("q_fifo");
        QueueClient q = helpers.queues();
        q.create(qn);
        q.push(qn, Map.of("v", 1));
        q.push(qn, Map.of("v", 2));
        assertEquals(2L, q.len(qn));
        List<Object> peek = q.peek(qn, 1);
        assertEquals(1, peek.size());
        List<Object> p1 = q.pop(qn, 1);
        assertEquals(1, p1.size());
        List<Object> p2 = q.pop(qn, 1);
        assertEquals(1, p2.size());
    }

    @Test void TestConformance_queues_empty_pop_returns_empty() {
        String qn = uniq("q_empty");
        helpers.queues().create(qn);
        List<Object> p = helpers.queues().pop(qn, 1);
        assertTrue(p.isEmpty());
    }

    @Test void TestConformance_queues_purge_resets_len() {
        String qn = uniq("q_purge");
        QueueClient q = helpers.queues();
        q.create(qn);
        q.push(qn, "a");
        q.push(qn, "b");
        q.purge(qn);
        assertEquals(0L, q.len(qn));
    }

    // ---- tx --------------------------------------------------------

    @Test void TestConformance_tx_commit_persists() {
        String col = uniq("tx_commit");
        helpers.documents().insert(col, Map.of("seed", true));  // ensure collection
        TxClient tx = helpers.tx();
        tx.begin();
        String rid = helpers.documents().insert(col, Map.of("x", 1)).rid();
        tx.commit();
        var got = helpers.documents().get(col, rid);
        assertNotNull(got);
    }

    @Test void TestConformance_tx_rollback_discards() {
        String col = uniq("tx_rb");
        helpers.documents().insert(col, Map.of("seed", true));  // ensure collection
        TxClient tx = helpers.tx();
        tx.begin();
        String rid = helpers.documents().insert(col, Map.of("x", 1)).rid();
        tx.rollback();
        assertThrows(HelperException.NotFound.class,
            () -> helpers.documents().get(col, rid));
    }

    // ---- errors ----------------------------------------------------

    @Test void TestConformance_errors_not_found_document_get() {
        String col = uniq("err_nf");
        helpers.documents().insert(col, Map.of("seed", true));  // ensure collection
        assertThrows(HelperException.NotFound.class,
            () -> helpers.documents().get(col, "no-such-rid"));
    }

    // ---- provisional wire ------------------------------------------

    @Test void TestConformance_wire_probabilistic_hll_round_trip() {
        String name = uniq("hll");
        conn.query("CREATE HLL " + name);
        conn.query("HLL ADD " + name + " 'user1' 'user2'");
        byte[] body = conn.query("HLL COUNT " + name);
        String s = new String(body, StandardCharsets.UTF_8);
        assertTrue(s.contains("count") || s.contains("cardinality"),
            "expected count or cardinality column: " + s);
    }

    // ---- spec version constant -------------------------------------

    @Test void TestConformance_meta_spec_version() {
        assertEquals("1.0", Helpers.HELPER_SPEC_VERSION);
    }

    // ---- support helpers ------------------------------------------

    private static List<String> redCommand(Path dataPath, int port) {
        String redBin = System.getenv("RED_BIN");
        List<String> cmd = new ArrayList<>();
        if (redBin != null && !redBin.isBlank()) {
            cmd.add(redBin);
        } else {
            cmd.add("cargo");
            cmd.add("run");
            cmd.add("--release");
            cmd.add("--bin");
            cmd.add("red");
            cmd.add("--");
        }
        cmd.add("server");
        cmd.add("--path");
        cmd.add(dataPath.toString());
        cmd.add("--bind");
        cmd.add("127.0.0.1:" + port);
        return cmd;
    }

    private static Conn waitForConnect(String uri, Duration timeout) throws Exception {
        long deadline = System.nanoTime() + timeout.toNanos();
        RuntimeException last = null;
        while (System.nanoTime() < deadline) {
            try {
                Conn c = Reddb.connect(uri);
                try {
                    c.ping();
                    return c;
                } catch (RuntimeException e) {
                    c.close();
                    last = e;
                }
            } catch (RuntimeException e) {
                last = e;
            }
            Thread.sleep(50);
        }
        IOException error = new IOException("server did not accept connections at " + uri);
        if (last != null) error.initCause(last);
        throw error;
    }

    private static int freePort() throws IOException {
        try (ServerSocket socket = new ServerSocket(0, 0, InetAddress.getByName("127.0.0.1"))) {
            return socket.getLocalPort();
        }
    }

    private static File findRepoRoot() {
        File f = new File("").getAbsoluteFile();
        while (f != null) {
            if (new File(f, "Cargo.toml").exists() && new File(f, "drivers").isDirectory()) {
                return f;
            }
            f = f.getParentFile();
        }
        fail("could not locate repo root with Cargo.toml + drivers/");
        return null;
    }
}
